src/main/java/org/elasticsearch/env/NodeEnvironment.java

```diff
@@ -771,4 +771,17 @@ private Path resolveCustomLocation(@IndexSettings Settings indexSettings, final
     public Path resolveCustomLocation(@IndexSettings Settings indexSettings, final ShardId shardId) {
         return resolveCustomLocation(indexSettings, shardId.index().name()).resolve(Integer.toString(shardId.id()));
     }
+
+    /**
+     * Returns the {@code NodePath.path} for this shard.
+     */
+    public static Path shardStatePathToDataPath(Path shardPath) {
```

Comment:
> this only works if the custom data path is under the node path right?

See above: because the shardStatePath is now always on the local path.data, this should work in the custom data path case too?